File: internal/web/static/exttools.js
// exttools.js — the External tools overlay: a READ-ONLY inventory of the
// configured [[tools.command]] blocks (every category, every frontend) and
// the catalog tools detected on this machine. Adding/editing stays in the
// TUI Settings wizard; this surface answers "what is configured, is it
// approved for this repo, and what could I add".
import { $, esc, getJSON } from "./core.js";
import { closeLayer, pushLayer } from "./layers.js";

let data = null; // last GET /api/exttools payload

async function openExtToolsView() {
  try {
    data = await getJSON("/api/exttools");
  } catch (e) {
    return;
  }
  renderExtTools();
  pushLayer("exttools", $("exttools"), {
    onKey: (e) => {
      if (e.key === "Escape") {
        closeLayer("exttools");
        e.preventDefault();
        return true;
      }
      return true; // read-only surface: swallow app hotkeys, nothing else
    },
  });
}

function cmdRowHTML(c) {
  const badges = [c.category, c.mode, c.per_file ? "per-file" : "", c.when_op, ...(c.frontends || [])]
    .filter(Boolean)
    .map((b) => `<span class="xbadge">${esc(b)}</span>`)
    .join("");
  const approval = c.approved
    ? `<span class="xok">approved</span>`
    : `<span class="snote">not yet approved</span>`;
  const problem = c.valid ? "" : `<div class="xproblem">${esc(c.problem)}</div>`;
  return `
    <div class="xcmd">
      <div class="srow"><span class="sval">${esc(c.name)}</span>${badges}${approval}</div>
      <div class="xcmdline">${esc(c.command)}</div>
      ${problem}
    </div>`;
}

function detRowHTML(d) {
  const tmpls = (d.templates || [])
    .map((t) => {
      const marks = [t.configured ? "configured ✓" : "not configured", t.opt_in ? "opt-in" : ""]
        .filter(Boolean)
        .join(" · ");
      return `<div class="srow xtmpl"><span class="xbadge">${esc(t.category)}</span><span class="sval">${esc(t.name)}</span><span class="snote">${esc(marks)}</span></div>`;
    })
    .join("");
  return `
    <div class="xdet">
      <div class="srow"><span class="sval">${esc(d.label)}</span><span class="snote">${esc(d.bin)}</span></div>
      ${tmpls}
    </div>`;
}

function renderExtTools() {
  if (!data) return;
  const cmds = (data.commands || []).map(cmdRowHTML).join("");
  const dets = (data.detected || []).map(detRowHTML).join("");
  $("exttools-box").innerHTML = `
    <h2>external tools</h2>
    <h3>configured commands</h3>
    ${cmds || '<div class="srow"><span class="snote">(none configured)</span></div>'}
    <h3>detected on this machine</h3>
    ${dets || '<div class="srow"><span class="snote">(no catalog tools found)</span></div>'}
    <div class="sfoot">read-only — add or edit in the TUI settings (,) → external tools; the wizard writes to ${esc(data.global_config_path || "the global config")} · a command runs only after you approve its full text, per repo · approvals are shared between the TUI and this page · esc closes</div>`;
}

$("exttools").addEventListener("click", (e) => {
  if (e.target === $("exttools")) closeLayer("exttools");
});

export { openExtToolsView };

File: internal/web/exttools.go
package web

import (
	"net/http"
	"os"
	"os/exec"

	"github.com/homeend/gigagit/internal/config"
	"github.com/homeend/gigagit/internal/exttool"
	"github.com/homeend/gigagit/internal/promptstate"
	"github.com/homeend/gigagit/internal/template"
)

// The external-tools VIEW: a read-only inventory of the configured
// [[tools.command]] blocks (every category, every frontend — including rows
// this web frontend itself would filter out at run time) plus which catalog
// tools are detected on this machine. Adding/editing stays in the TUI
// Settings wizard for now; this surface answers "what is configured, is it
// approved, and what could I add".

type extToolCmdRow struct {
	Category  string   `json:"category"`
	Name      string   `json:"name"`
	Mode      string   `json:"mode"`
	PerFile   bool     `json:"per_file"`
	WhenOp    string   `json:"when_op"`
	Frontends []string `json:"frontends"`
	Command   string   `json:"command"`
	// Valid reports the same structural checks the run-time lanes apply
	// before offering a command; Problem carries the first failure's text so
	// a misconfigured block is diagnosable from the browser.
	Valid   bool   `json:"valid"`
	Problem string `json:"problem"`
	// Approved: this repo has already approved this exact command text
	// (promptstate CommandHash — the store the TUI and the web lanes share).
	Approved bool `json:"approved"`
}

type extToolTemplateRow struct {
	Category   string `json:"category"`
	Name       string `json:"name"`
	OptIn      bool   `json:"opt_in"`
	Configured bool   `json:"configured"`
}

type extToolDetectedRow struct {
	ID        string               `json:"id"`
	Label     string               `json:"label"`
	Bin       string               `json:"bin"`
	Templates []extToolTemplateRow `json:"templates"`
}

// detections runs the catalog probe (or the test seam).
func (s *Server) detections() []exttool.Detection {
	if s.detectTools != nil {
		return s.detectTools()
	}
	home, _ := os.UserHomeDir()
	return exttool.Detect(exec.LookPath, os.Stat, home)
}

func (s *Server) handleExtTools(w http.ResponseWriter, r *http.Request) {
	svc := s.service()
	cfg, err := s.effectiveConfig(r.Context(), svc)
	if err != nil {
		writeErr(w, http.StatusInternalServerError, err)
		return
	}
	var approved map[string]bool
	if store := s.promptStore(); store != nil {
		approved = store.ApprovedToolCommands(s.toolRepoKey(r.Context(), svc))
	}
	cmds := make([]extToolCmdRow, 0, len(cfg.Tools.Command))
	have := map[string]bool{}
	for _, tc := range cfg.Tools.Command {
		have[tc.Key()] = true
		row := extToolCmdRow{
			Category: tc.Category, Name: tc.Name, Mode: tc.Mode, PerFile: tc.PerFile,
			WhenOp: tc.WhenOp, Frontends: tc.Frontends, Command: tc.Command,
			Valid:    true,
			Approved: approved[promptstate.CommandHash(tc.Command)],
		}
		if verr := config.ValidateToolCommand(tc); verr != nil {
			row.Valid, row.Problem = false, verr.Error()
		} else if terr := template.ValidateCommandTokens(tc.Command, tc.PerFile); terr != nil {
			row.Valid, row.Problem = false, terr.Error()
		}
		cmds = append(cmds, row)
	}
	dets := make([]extToolDetectedRow, 0, 4)
	for _, det := range s.detections() {
		row := extToolDetectedRow{ID: det.Tool.ID, Label: det.Tool.Label, Bin: det.Bin}
		for _, ct := range det.Tool.Commands {
			row.Templates = append(row.Templates, extToolTemplateRow{
				Category: string(ct.Category), Name: ct.Name, OptIn: ct.OptIn,
				Configured: have[string(ct.Category)+"\x00"+ct.Name],
			})
		}
		dets = append(dets, row)
	}
	writeJSON(w, map[string]any{
		"commands":           cmds,
		"detected":           dets,
		"global_config_path": config.DefaultGlobalPath(),
	})
}
